File: source/49.字母异位词分组.cpp
/*
 * @lc app=leetcode.cn id=49 lang=cpp
 *
 * [49] 字母异位词分组
 */

// @lc code=start
class Solution {
 public:
  vector<vector<string>> groupAnagrams(vector<string>& strs) {
    vector<vector<string>> res;
    unordered_map<string, int> hashmap;

    int index = 0;
    for (string s : strs) {
      string t = s;
      sort(t.begin(), t.end());
      if (hashmap.find(t) != hashmap.end()) {
        res[hashmap[t]].push_back(s);
      } else {
        hashmap[t] = index++;
        res.push_back({s});
      }
    }

    return res;
  }
};
// @lc code=end

class Solution {
 public:
  vector<vector<string>> groupAnagrams(vector<string>& strs) {
    vector<vector<string>> res;
    unordered_map<string, vector<string>> hashmap;

    for (string s : strs) {
      string t = s;
      sort(t.begin(), t.end());
      hashmap[t].push_back(s);
    }

    for (auto kv : hashmap) {
      res.push_back(kv.second);
    }

    return res;
  }
};

File: source/49.字母异位词分组.cs
/*
 * @lc app=leetcode.cn id=49 lang=csharp
 *
 * [49] 字母异位词分组
 */

// @lc code=start
public class Solution {

    public IList<IList<string>> GroupAnagrams(string[] strs) {
        var result = new List<IList<string>>();
        if (strs.Length < 1) return result;

        var dict = new Dictionary<string, List<string>>();
        foreach(var s in strs) {
            var k = string.Concat(s.OrderBy(c => c));
            if (dict.ContainsKey(k))
                dict[k].Add(s);
            else
                dict[k] = new List<string>{s};
        }

        foreach(var kv in dict) {
            result.Add(kv.Value);
        }

        return result;
    }
}
// @lc code=end

